File: tests/tt_eager/python_api_testing/unit_testing/misc/test_sum.py
# SPDX-FileCopyrightText: © 2023-24 Tenstorrent Inc.

# SPDX-License-Identifier: Apache-2.0

import torch
import pytest
import ttnn


@pytest.mark.parametrize(
    "shape_dim",
    (
        ((1, 1, 32, 32), 3),
        ((1, 1, 32, 32), 2),
        ((32, 32, 32, 32), 1),
        ((32, 32, 32, 32), 0),
    ),  # single tile
)
def test_sum_for_dim_hw(device, use_program_cache, shape_dim):
    shape, dim = shape_dim
    torch.manual_seed(0)

    N = shape[0]
    C = shape[1]
    H = shape[2]
    W = shape[3]

    input_shape = (N, C, H, W)
    x = 1.0 + torch.arange(0, N * C * H * W).reshape(input_shape).bfloat16()

    value = x.sum(dim=dim, keepdim=True)[0, 0, 0, 0]
    # print(f"x.sum = {value}")

    dev_x = ttnn.Tensor(x, ttnn.DataType.BFLOAT16).to(ttnn.Layout.TILE).to(device)
    tt_npu = ttnn.sum(dev_x, dim)
    tt_dev = tt_npu.cpu().to(ttnn.Layout.ROW_MAJOR).to_torch()
    assert torch.equal(tt_dev[0, 0, 0, 0], torch.Tensor([value]).bfloat16()[0])


@pytest.mark.parametrize(
    "shape",
    (
        (1, 1, 32, 32),
        (1, 1, 32, 32),
        (32, 32, 32, 32),
        (32, 32, 32, 32),
    ),  # single tile
)
def test_sum_global(device, use_program_cache, shape):
    torch.manual_seed(0)

    N = shape[0]
    C = shape[1]
    H = shape[2]
    W = shape[3]

    input_shape = (N, C, H, W)
    x = 1.0 + torch.ones(input_shape).bfloat16()

    value = x.sum()

    dev_x = ttnn.Tensor(x, ttnn.DataType.BFLOAT16).to(ttnn.Layout.TILE).to(device)
    tt_npu = ttnn.sum(dev_x)
    tt_dev = tt_npu.cpu().to(ttnn.Layout.ROW_MAJOR).to_torch()
    assert torch.equal(tt_dev[0, 0, 0, 0].bfloat16(), torch.Tensor([value]).bfloat16()[0])

File: tests/ttnn/unit_tests/gtests/test_sum.cpp
// SPDX-FileCopyrightText: © 2026 Tenstorrent AI ULC
//
// SPDX-License-Identifier: Apache-2.0

#include <fmt/base.h>
#include <gtest/gtest.h>
#include <cstdint>
#include <array>
#include <memory>
#include <optional>

#include <tt_stl/assert.hpp>
#include <tt-metalium/bfloat16.hpp>
#include <tt-metalium/shape.hpp>
#include "ttnn/decorators.hpp"
#include "ttnn/device.hpp"
#include "ttnn/operations/core/core.hpp"
#include "ttnn/operations/creation/creation.hpp"
#include "ttnn/operations/eltwise/binary/binary.hpp"
#include "ttnn/operations/functions.hpp"
#include "ttnn/operations/reduction/generic/generic_reductions.hpp"
#include "ttnn/tensor/shape/shape.hpp"
#include "ttnn/tensor/types.hpp"
#include "ttnn/types.hpp"
#include "ttnn_test_fixtures.hpp"

namespace ttnn::operations::binary::test {

struct SumTensorLastDimParam {
    int32_t dim;
    uint32_t h;
    uint32_t w;
};

class SumTensorLastDimFixture : public TTNNFixtureWithSuiteDevice<SumTensorLastDimFixture>,
                                public testing::WithParamInterface<SumTensorLastDimParam> {};

TEST_P(SumTensorLastDimFixture, SumTensorCorrectly) {
    auto param = GetParam();
    auto& device = *device_;
    std::array<uint32_t, 2> dimensions = {param.h, param.w};
    ttnn::Shape shape(dimensions);
    std::array<uint32_t, 2> reduced_dimensions = {param.h, 1};
    ttnn::Shape reduced_shape(reduced_dimensions);

    {
        const auto input_tensor = ttnn::ones(shape, DataType::BFLOAT16, ttnn::TILE_LAYOUT, device);
        const auto output_tensor = ttnn::sum(
            input_tensor,
            param.dim,
            true);  // keepdim true allows output of reduce to be returned directly without reshape on device call
        TT_FATAL(
            output_tensor.logical_shape() == reduced_shape,
            "Shapes are not equal output tensor shape {} vs reduced shape {}",
            output_tensor.logical_shape(),
            reduced_shape);
        auto output_vector = output_tensor.to_vector<bfloat16>();
        float expected = (float)param.w;
        for (uint32_t i = 0; i < param.h; i++) {
            float value = output_vector[i];
            TT_FATAL(value == expected, "{} != {} @ {}", value, expected, i);
        }
    }
}

INSTANTIATE_TEST_SUITE_P(
    SumTensorLastDimTests, SumTensorLastDimFixture, ::testing::Values(SumTensorLastDimParam{-1, 32, 64}));

}  // namespace ttnn::operations::binary::test
